File: internal/bitrope/bitrope.go
package bitrope

// Append-only packed bit rope with fixed-size blocks.
//
// Primary goal: predictable allocations + cheap appends + cheap window reads.
// This is a pragmatic building block for unbounded boolean streams.

import "fmt"

type Cursor struct {
	Block int
	Off   int // bit offset within block
}

type Rope struct {
	BlockBits int
	blocks    [][]uint64
	nBitsLast int
	lenBits   uint64
}

func New(blockBits int) *Rope {
	if blockBits <= 0 {
		blockBits = 1 << 16 // 65536 bits
	}
	// round up to multiple of 64
	if blockBits%64 != 0 {
		blockBits = ((blockBits + 63) / 64) * 64
	}
	return &Rope{BlockBits: blockBits}
}

func (r *Rope) LenBits() uint64 { return r.lenBits }

func (r *Rope) Blocks() (numBlocks int, blockBits int) { return len(r.blocks), r.BlockBits }

func (r *Rope) ensureBlock() {
	if len(r.blocks) == 0 || r.nBitsLast >= r.BlockBits {
		nWords := r.BlockBits / 64
		r.blocks = append(r.blocks, make([]uint64, nWords))
		r.nBitsLast = 0
	}
}

// AppendBit appends a single bit and returns its cursor.
func (r *Rope) AppendBit(b uint8) Cursor {
	b &= 1
	r.ensureBlock()
	bi := len(r.blocks) - 1
	off := r.nBitsLast
	wi := off >> 6
	bit := off & 63
	if b == 1 {
		r.blocks[bi][wi] |= 1 << bit
	}
	r.nBitsLast++
	r.lenBits++
	return Cursor{Block: bi, Off: off}
}

// Get returns the bit at absolute position i; out-of-range reads return 0.
func (r *Rope) Get(i uint64) uint8 {
	if i >= r.lenBits {
		return 0
	}
	bi := int(i / uint64(r.BlockBits))
	off := int(i) - bi*r.BlockBits
	wi := off >> 6
	bit := off & 63
	return uint8((r.blocks[bi][wi] >> bit) & 1)
}

// ReadBits returns n bits starting at start. Out-of-range bits are 0.
func (r *Rope) ReadBits(start uint64, n int) []uint8 {
	if n <= 0 {
		return nil
	}
	out := make([]uint8, n)
	for j := 0; j < n; j++ {
		out[j] = r.Get(start + uint64(j))
	}
	return out
}

// ReadU64Window reads up to 64 bits into an LSB-first uint64.
func (r *Rope) ReadU64Window(start uint64, n int) uint64 {
	if n <= 0 {
		return 0
	}
	if n > 64 {
		n = 64
	}
	var x uint64
	for j := 0; j < n; j++ {
		x |= (uint64(r.Get(start+uint64(j))) & 1) << uint(j)
	}
	return x
}

func (r *Rope) String() string {
	return fmt.Sprintf("Rope{lenBits=%d blocks=%d blockBits=%d}", r.lenBits, len(r.blocks), r.BlockBits)
}

File: scripts/python_ref/bitrope.py
#!/usr/bin/env python3
"""bitrope.py — append-only immutable-ish bit rope (Python reference)

Goal: support unbounded boolean streams with:
- bounded allocation (fixed-size blocks)
- cheap append
- cheap window reads for analysis/transforms

This is a reference implementation for experiments and dashboards.
It is not meant to be the fastest possible Python.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple


@dataclass(frozen=True)
class Cursor:
    block: int
    off: int  # bit offset inside block


class BitRope:
    def __init__(self, block_bits: int = 1 << 16):
        if block_bits <= 0:
            block_bits = 1 << 16
        if block_bits % 64 != 0:
            # keep packing simple
            block_bits = ((block_bits + 63) // 64) * 64
        self.block_bits = block_bits
        self._blocks: List[List[int]] = []  # each block: list of uint64 words
        self._n_bits_in_last = 0
        self._len_bits = 0

    def __len__(self) -> int:
        return self._len_bits

    def _ensure_block(self):
        if not self._blocks or self._n_bits_in_last >= self.block_bits:
            n_words = self.block_bits // 64
            self._blocks.append([0] * n_words)
            self._n_bits_in_last = 0

    def append_bit(self, b: int) -> Cursor:
        b &= 1
        self._ensure_block()
        bi = len(self._blocks) - 1
        off = self._n_bits_in_last
        wi = off >> 6
        bit = off & 63
        if b:
            self._blocks[bi][wi] |= (1 << bit)
        self._n_bits_in_last += 1
        self._len_bits += 1
        return Cursor(bi, off)

    def get(self, i: int) -> int:
        if i < 0 or i >= self._len_bits:
            return 0
        bi = i // self.block_bits
        off = i - bi * self.block_bits
        wi = off >> 6
        bit = off & 63
        return (self._blocks[bi][wi] >> bit) & 1

    def read_bits(self, start: int, n: int) -> List[int]:
        """Read n bits starting at start (out of range -> 0)."""
        return [self.get(start + j) for j in range(max(0, n))]

    def read_u64_window(self, start: int, n: int) -> int:
        """Read up to 64 bits into LSB-first integer."""
        n = min(64, max(0, n))
        x = 0
        for j in range(n):
            x |= (self.get(start + j) & 1) << j
        return x

    def blocks(self) -> Tuple[int, int]:
        """(num_blocks, block_bits)."""
        return (len(self._blocks), self.block_bits)
